### PR TITLE
Add CORS support to APIGW

### PR DESCRIPTION
**Pull request by patoncrispy**
##### Status:
Ready

##### Description:
Have enabled CORS to be configured on a http event. The preflight (OPTIONS) request is configured automatically and does not need to be specified. Header and Origins settings can be configured from the `serverless.yml` file. Have updated the readme accordingly. Tests have been written to validate the configuration.

### CODE REVIEW

**Comment by patoncrispy** [comment]
@pmuens @flomotlik @eahefnawy I have closed https://github.com/serverless/serverless/pull/1703 and opened this bad boy up. The rebasing was a bit scary, however it all seems OK. Would someone be able to double-check this to make sure I haven't screwed up something else in the process, please? 😰

**Comment by flomotlik** [comment]
@patoncrispy looking into it today

File: lib/plugins/aws/deploy/compile/events/apiGateway/README.md [hunk 1]
```diff
@@ -157,3 +157,40 @@ That wouldn't be required if you hadn't set the `private` property to `true`.
...
+              - '*'
+            headers:
+              - Content-Type
+              - X-Amz-Date
+              - Authorization
+              - X-Api-Key
+              - X-Amz-Security-Token
+```
+
+This is example is the default setting and is exactly the same as the previous example. The `Access-Control-Allow-Methods` header is set automatically, based on the endpoints specified in your service configuration with CORS enabled.
```
Comment (flomotlik) [inline lib/plugins/aws/deploy/compile/events/apiGateway/README.md]: A typo here: "This is example is"

Comment (patoncrispy) [inline lib/plugins/aws/deploy/compile/events/apiGateway/README.md]: :) I'll sort this out this evening.

**Comment by ac360** [comment]
I am so excited for this feature.  @patoncrispy, you are truly great.

**Comment by patoncrispy** [comment]
Thanks, @ac360! :blush: This was my first open source contribution that I've ever made, and @flomotlik @pmuens and @eahefnawy have made this so incredibly enjoyable and easy. Thank you all! I want to work on some more contributions and help you guys out, but I'm currently working on the Serverless Slack chatbot competition AWS are running. As soon as that's over, I'll be right back! :smiley:

**Comment by flomotlik** [comment]
@patoncrispy it looks like there are some conflicts now in the PR. Could you merge the master into your PR, then we can merge this. Sorry for the additional work and delay in this! Great stuff

**Comment by patoncrispy** [comment]
@flomotlik ah nuts! :smile:  Will sort this out later. it worked fine on my machine! :laughing:

**Comment by patoncrispy** [comment]
@flomotlik I've fixed everything. Updated to the latest master branch.

**Comment by flomotlik** [comment]
@patoncrispy looks like one of your merges took out the tests? So the PR is imho ready to be merged if you could readd the tests again from the git history. Happy to help.

**Comment by flomotlik** [comment]
Nope its good right now. Looks like the core tests are back, merging now

File: lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js [hunk 2]
```diff
@@ -21,13 +22,63 @@ module.exports = {
...
+              cors.methods = [];
+              if (cors.headers && !Array.isArray(cors.headers)) {
+                const errorMessage = [
+                  'CORS header values must be provided as an array.',
+                  ' Please check the docs for more info.',
+                ].join('');
+                throw new this.serverless.classes
+                .Error(errorMessage);
+              } else {
+                cors.headers = headers;
```
Comment (yonahforst) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js]: isn't this overwriting custom headers that may have been specified in the YML?

Comment (patoncrispy) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js]: No, this is purely setting the CORS configuration. AFAIK, it isn't possible to set CORS headers anywhere else in the Serverless configuration. @pmuens, can you verify this? Of course, if you are experiencing problems, I'm sure the Serverless team would want you to raise an Issue. :)

Comment (pmuens) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js]: Yes, right now you can only set CORS headers in the `http` event definition. #1968 will introduce a way to setup your own response headers if you'd like to have more control over it.

Comment (yonahforst) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js]: I should have been more specific 😅  I was referring to L61. I think it was overwriting the headers in set my http event definition. I opened #1984 to track.

Comment (pmuens) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js]: Great. Thanks for that 👍

**Pull request merged**